src/OAuth2/Controller/AuthorizeController.php

```diff
@@ -334,13 +349,50 @@ public function validateAuthorizeRequest(RequestInterface $request, ResponseInte
             return false;
         }
 
+        $code_challenge = $request->query('code_challenge');
+        $code_challenge_method = $request->query('code_challenge_method');
+        if ($code_challenge) {
+            if (preg_match('/^[A-Za-z0-9-._~]{43,128}$/', $code_challenge) !== 1) {
+                $response->setError(400, 'invalid_code_challenge', 'The PKCE code challenge supplied is invalid');
+
+                return false;
+            }
+
+            $supported_challenge_methods = $this->config['supported_code_challenge_methods'] ?? ['plain', 'S256'];
+            if (!$code_challenge_method) {
+                $response->setError(400, 'missing_code_challenge_method', 'This application requires you specify a PKCE code challenge method. Supported methods: ' . implode(', ', $supported_challenge_methods)
+                );
+
+                return false;
+            }
+            if (!in_array($code_challenge_method, $supported_challenge_methods, true)) {
+                $response->setError(400, 'invalid_code_challenge_method', 'The PKCE code challenge method supplied is invalid. Supported methods: ' . implode(', ', $supported_challenge_methods)
+                );
+
+                return false;
+            }
+        } else {
+            $isPublic = false;
+            if ($this->clientStorage instanceof ClientCredentialsInterface) {
+                $isPublic = $this->clientStorage->isPublicClient($client_id);
```

Comment:
ℹ️ not sure why there are two interfaces, but the property is only typed to the higher level one which doesn't have the method we need.